django/db/backends/__init__.py

```diff
@@ -823,6 +823,15 @@ def cache_key_culling_sql(self):
         """
         return "SELECT cache_key FROM %s ORDER BY cache_key LIMIT 1 OFFSET %%s"
 
+    def case_cast_sql(self, db_type, internal_type):
```

Comment:
Michal and I discussed this on IRC and I don't think such a generic function is going to help here. This is a very narrow use case (implicit casting isn't working for certain fields in case expressions), and it needs to be special cased for certain fields. This isn't an Integer->Float kind of situation.

I'm happy enough with the method to stay. If we can generalise later then that will be OK, but I don't want this to hold up the merge.